cmd/util/ledger/migrations/cadence_value_diff.go

```diff
@@ -96,81 +100,130 @@ func NewCadenceValueDiffReporter(
 
-func (dr *CadenceValueDiffReporter) DiffStates(oldRegs, newRegs registers.Registers, domains []string) {
+	newStorage := newReadonlyStorage(newRegs)
 
-	// Create all the runtime components we need for comparing Cadence values.
-	oldRuntime, err := dr.newStorageRuntime(oldRegs)
+	var loadAtreeStorageGroup errgroup.Group
+
+	loadAtreeStorageGroup.Go(func() (err error) {
+		return loadAtreeSlabsInStorage(oldStorage, oldRegs, dr.nWorkers)
+	})
+
+	err := loadAtreeSlabsInStorage(newStorage, newRegs, dr.nWorkers)
 	if err != nil {
 		dr.reportWriter.Write(
 			diffError{
 				Address: dr.address.Hex(),
 				Kind:    diffErrorKindString[abortErrorKind],
-				Msg:     fmt.Sprintf("failed to create runtime for old registers: %s", err),
+				Msg:     fmt.Sprintf("failed to preload new atree registers: %s", err),
 			})
 		return
 	}
 
-	newRuntime, err := dr.newStorageRuntime(newRegs)
-	if err != nil {
+	// Wait for old registers to be loaded in storage.
+	if err := loadAtreeStorageGroup.Wait(); err != nil {
 		dr.reportWriter.Write(
 			diffError{
 				Address: dr.address.Hex(),
 				Kind:    diffErrorKindString[abortErrorKind],
-				Msg:     fmt.Sprintf("failed to create runtime with new registers: %s", err),
+				Msg:     fmt.Sprintf("failed to preload old atree registers: %s", err),
 			})
 		return
 	}
 
-	err = loadAtreeSlabsInStorage(oldRuntime.Storage, oldRegs, dr.nWorkers)
+	if oldRegs.Count() > minLargeAccountRegisterCount {
+		// Add concurrency to diff domains
+		var g errgroup.Group
+
+		// NOTE: preload storage map in the same goroutine
+		for _, domain := range domains {
+			_ = oldStorage.GetStorageMap(dr.address, domain, false)
+			_ = newStorage.GetStorageMap(dr.address, domain, false)
+		}
+
+		// Create goroutine to diff storage domain
+		g.Go(func() (err error) {
+			oldRuntime, err := newReadonlyStorageRuntimeWithStorage(oldStorage, oldRegs.Count())
+			if err != nil {
+				return fmt.Errorf("failed to create runtime for old registers: %s", err)
+			}
+
+			newRuntime, err := newReadonlyStorageRuntimeWithStorage(newStorage, newRegs.Count())
+			if err != nil {
+				return fmt.Errorf("failed to create runtime for new registers: %s", err)
+			}
+
+			dr.diffStorageDomain(oldRuntime, newRuntime, common.PathDomainStorage.Identifier())
+			return nil
+		})
+
+		// Create goroutine to diff other domains
+		g.Go(func() (err error) {
+			oldRuntime, err := newReadonlyStorageRuntimeWithStorage(oldStorage, oldRegs.Count())
+			if err != nil {
+				return fmt.Errorf("failed to create runtime for old registers: %s", err)
+			}
+
+			newRuntime, err := newReadonlyStorageRuntimeWithStorage(newStorage, oldRegs.Count())
+			if err != nil {
+				return fmt.Errorf("failed to create runtime for new registers: %s", err)
+			}
+
+			for _, domain := range domains {
+				if domain != common.PathDomainStorage.Identifier() {
+					dr.diffStorageDomain(oldRuntime, newRuntime, domain)
+				}
+			}
+			return nil
+		})
```

Comment:
Is the idea behind the split of the storage domain and the other domains that the storage domain usually has more data stored than the other domains?

Reply:
> Is the idea behind the split of the storage domain and the other domains that the storage domain usually has more data stored than the other domains?

Yes.  Storage domain usually has more data stored than other domains.  

The separate optimization PR I mentioned this morning is to split storage domain but not sure if it would work.